pymeasure/instruments/attocube/adapters.py

```diff
@@ -0,0 +1,104 @@
+# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+# THE SOFTWARE.
+#
+
+import time
+
+from pymeasure.adapters import TelnetAdapter
+
+
+class AttocubeConsoleAdapter(TelnetAdapter):
+    """ Adapter class for connecting to the Attocube Standard Console. This
+    console is a Telnet prompt with password authentication.
+
+    :param host: host address of the instrument
+    :param port: TCPIP port
+    :param passwd: password required to open the connection
+    :param kwargs: Any valid key-word argument for TelnetAdapter
+    """
+    def __init__(self, host, port, passwd, **kwargs):
+        self.terminator = '\r\n'
+        super(AttocubeConsoleAdapter, self).__init__(host, port, **kwargs)
+        # clear the initial messages from the controller
+        time.sleep(self.query_delay)
+        super(AttocubeConsoleAdapter, self).read()
+        # send password and check authorization
+        self.write(passwd, checkAck=False)
+        time.sleep(self.query_delay)
+        ret = super(AttocubeConsoleAdapter, self).read()
+        authmsg = ret.split(self.terminator)[1]
+        if authmsg != 'Authorization success':
+            raise Exception("Attocube authorization failed ('%s')" % authmsg)
+        # switch console echo off
+        self.ask('echo off')
+
+    def check_acknowledgement(self, reply, msg=""):
+        """ Reads the reply of the instrument until the termination sequence.
+        The returned string is check to be 'OK' and otherwise a ValueError is
+        raised.
+
+        :param msg: optional message for the eventual error
+        :returns: String ASCII response of the instrument.
+        """
+        if reply.strip() != 'OK':
+            if msg == "":  # clear buffer
+                msg = reply.strip()
+                super(AttocubeConsoleAdapter, self).read()
+            raise ValueError("AttocubeConsoleAdapter: Error after command '%s'"
+                             " with message '%s'" % (self.lastcommand, msg))
+
+    def read(self):
+        """ Reads a reply of the instrument which consists of two or more
+        lines. The first ones are the reply to the command while the last one
+        is 'OK' or 'ERROR' to indicate any problem. In case the reply is not OK
+        a ValueError is raised.
+
+        :returns: String ASCII response of the instrument.
+        """
+        raw = super(AttocubeConsoleAdapter, self).read().strip()
+        lines = raw.split('\n')  # line endings inconsistent '\n', or '\r\n'
+        ret = '\n'.join(line.strip() for line in lines[:-1])
+        self.check_acknowledgement(lines[-1], ret)
+        return ret
+
+    def write(self, command, checkAck=True):
+        """ Writes a command to the instrument
+
+        :param command: command string to be sent to the instrument
+        """
+        self.lastcommand = command
+        super(AttocubeConsoleAdapter, self).write(command + self.terminator)
+        if checkAck:
```

Comment:
watch out here with API structure, your `write` now effectively is an `ask` (write-read cycle), which feels weird/inconsistent with the other adapters.

Reply:
The device responds to every command (also a set command) by replying 'OK' or 'ERROR'. Thats why in the write I also read back the reply of the instrument. I understand this could be handled by the check_errors method discussed above. And this is the reason why ask method uses `checkAck = False`. For a command with a reply value the 'OK' is received after the return value. The return can, however, consist of multiple lines which makes the read complicated since one does not know how far to read to not read the 'OK' which would need to be again handled by  `check_error`. advise is welcome

Reply:
> the return can, however, consist of multiple lines which makes the read complicated since one does not know how far to read to not read the 'OK' which would need to be again handled by check_error. advise is welcome

OK in that case check_error would indeed be weird to use.